src/Types/Infer/App.hs

```diff
@@ -0,0 +1,224 @@
+--
+--  * The first pass is what we call a 'quick look'. Here, we try to
+--  infer polymorphic type instantiations for some of the function's
+--  arguments. This pass only considers "simple" arguments: Variables
+--  (for now).
+--
+--  * The second pass is the real use of 'checks' that we've always done
+--  for arguments.
+inferApps :: forall m. MonadInfer Typed m => Expr Desugared -> Maybe (Type Typed) -> m (Expr Typed, Type Typed)
+inferApps exp expected =
+  do
+    let ((ExprArg function, _):arguments) = reverse $ spine exp
+    (function, function_t) <- infer' function
+    function_t <- refresh function_t
+
+#ifdef TRACE_TC
+    traceM ("function type: " ++ displayS (pretty function_t))
+#endif
+
+    ((ql_sub, quantifiers), cs) <-
+      censor (const mempty) . listen $
+        quickLook function_t =<< for arguments (\a@(x, y) -> (x, y,) <$> inferQL a)
+
+    quantifiers <- pure (applyQ ql_sub quantifiers)
+
+    let fake_ty_con = TyCon (TgInternal T.empty)
+    for_ (Map.toList ql_sub) $ \(var, tau) ->
+      -- Here we need these to be guarded, so they need to be in the RHS
+      -- of a type application. Since we don't have a suitable type, we
+      -- use an internal TyCon with an invisible name. (Evil!)
+      unify (becauseExp exp) (TyApp fake_ty_con (TyVar var)) (TyApp fake_ty_con tau)
+
+    tell cs
+
+    r_ql_sub <- case expected of
+      Just tau@(TyApps (TyCon _) (_:_)) | result@(TyApps (TyCon _) (_:_)) <- getQuantR quantifiers -> do
+-- Pushing down the result type into the quick look substitution
+-- is only sound if both are /guarded/. Do note that missing a
+-- substitution here isn't unsound, it'll just lead to wonky inference.
+        _ <- subsumes (becauseExp exp) result (apply ql_sub tau)
+        pure $ fromMaybe mempty (unifyPure result tau)
+      _ -> pure mempty
+
+#ifdef TRACE_TC
+    traceM ("Quick Look results: " ++ show (pretty <$> (ql_sub <> r_ql_sub), pretty quantifiers))
+#endif
+
+    (arg_ks, result) <-
+      checkArguments arguments (applyQ r_ql_sub quantifiers)
+
+#ifdef TRACE_TC
+    traceM ("resulting type: " ++ displayS (pretty result))
+#endif
+
+    wrap <- case expected of
+      Just tau -> do
+#ifdef TRACE_TC
+        traceM ("QL expected result: " ++ displayS (pretty tau))
+#endif
+        wrap <- subsumes (becauseExp exp) result tau
+        pure $ \ex -> ExprWrapper wrap ex (annotation ex, tau)
+      Nothing -> pure id
+
+    pure (wrap (foldr (.) id (reverse arg_ks) function), result)
+  where
+    spine ex@(App fn arg _) =
+      let sp = spine fn
+       in (ExprArg arg, BecauseOf ex):sp
+    spine ex@(Vta fn arg _) =
+      let sp = spine fn
+       in (TypeArg arg, BecauseOf ex):sp
+    spine ex = [(ExprArg ex, BecauseOf ex)]
+
+    checkArguments ((ExprArg arg, _):as) (Quant tau dom cod inst_cont qs) =
+      case dom of
+        Anon dom -> do
+          x <- check arg dom
+
+          let cont ex = App ex x (annotation ex, cod)
```

Comment:
`annotation ex <> annotation x` I think (and similarly in the `TyApp` case. Currently this respans applications to contain the function name, which makes a couple of errors messages more confusing.

Reply:
done, too